### DISCUSSION TITLE
Could we have property of nullable struct return ref

### DISCUSSION THREAD

**Discussion by Thaina** [dotnet/csharplang#762] [General]
If it not be breaking change I wish `Value` property of nullable struct should return reference. But if it is I think we should have another property that would return reference of the struct in the nullable directly

Something like

```C#
Vector3? v = GetNullablePosition(name);
ref Vector3 rv = v.Reference; // throw error for null, like .Value

Console.WriteLine(v?.X); // 0.0f
rv.X += 10;
Console.WriteLine(v?.X); // 10.0f
```

**Comment by yaakov-h** [comment]
Shouldn't this be a dotnet/corefx or dotnet/standard issue?

**Comment by benaadams** [comment]
I assume is currently a language issue as you can't have null `ref`s set by C# https://github.com/dotnet/csharplang/issues/497

**Comment by yaakov-h** [comment]
You wouldn’t need to, if it threw an exception as described when HasValue is false.

**Comment by svick** [comment]
I don't think such `Reference` property would be allowed under current `ref` returns rules. Specifically, [`ref`-returning methods in structs are not allowed to return `this` or their own instance fields](https://github.com/dotnet/csharplang/blob/master/meetings/2015/LDM-2015-09-01.md#this-in-structs).

For example, [this code](https://sharplab.io/#v2:D4ZwLgTgrgxmAEBZAngOSgGwwQwEYYFMAeAFQD4BYAKAG9qBIE+AN2wygIG5qGICAzeEwBKAgnwB2MAvAC8ZeH0Gt2XagF8gA===):

```c#
struct MyNullable<T>
{
	T value;

	ref T Reference => ref value;
}
```

produces:

> error CS8170: Struct members cannot return 'this' or other instance members by reference

**Comment by sharwell** [comment]
`StrongBox<T>` and `ValueTuple<T>` solved this by using public fields instead of properties.

**Comment by yaakov-h** [comment]
Although if you used a field here then you can’t throw an exception if it’s null.

**Comment by Thaina** [comment]
@svick Thanks. This is very sad. There are many existing function return nullable struct or logic using array of nullable and it has no easy way to edit internal value without copy

**Comment by Thaina** [comment]
@sharwell `StrongBox<T>` is class so it need to instantiate and defeat the purpose of being struct

`ValueTuple<T>` `(bool,T)` is more likely but it can't use `??` or `?.` or protect the value from accessing when it being null
